velox/type/tests/OpaqueCustomTypesTest.cpp

```diff
@@ -0,0 +1,100 @@
+/*
+ * Copyright (c) Facebook, Inc. and its affiliates.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <gtest/gtest.h>
+
+#include "velox/common/base/tests/GTestUtils.h"
+#include "velox/type/OpaqueCustomTypes.h"
+
+using namespace facebook;
+using namespace facebook::velox;
+
+namespace {
+
+class OpaqueCustomTypeTest : public testing::Test {
+ public:
+  OpaqueCustomTypeTest()
+      : testCustomType_(TestCustomTypeRegister::VeloxType::get()) {}
+
+ protected:
+  struct TestCustomType {};
+  static constexpr char testCustomTypeName[] = "my_test_custom_type";
+  using TestCustomTypeRegister =
+      OpaqueCustomTypeRegister<TestCustomType, testCustomTypeName>;
+
+  TestCustomTypeRegister::VeloxTypePtr testCustomType_;
+
+  void TearDown() override {
+    TestCustomTypeRegister::unregisterType();
+    ASSERT_EQ(getCustomType(testCustomTypeName, {}), nullptr);
+    VELOX_ASSERT_THROW(
+        testCustomType_->getSerializeFunc(),
+        "No serialization function registered for my_test_custom_type");
+    VELOX_ASSERT_THROW(
+        testCustomType_->getDeserializeFunc(),
+        "No deserialization function registered for my_test_custom_type");
+  }
+
+  void testTypeSerde() {
+    // Test serialization/deserialization.
+    auto serializedCustomType = testCustomType_->serialize();
+    EXPECT_EQ(testCustomType_->name(), serializedCustomType["opaque"]);
+    auto deserializedType = Type::create(serializedCustomType);
+
+    // TODO: Ideally this should be EXPECT_EQ() but it doesn't work
+    // because Type::create() currently does not return the singleton instance.
+    // EXPECT_EQ(testCustomType, deserializedType);
+
+    EXPECT_EQ(testCustomType_->serialize(), deserializedType->serialize());
+  }
+};
+
+TEST_F(OpaqueCustomTypeTest, registerWithoutSerde) {
+  // Register type without serialization/deserialization and build a TypePtr.
+  TestCustomTypeRegister::registerType();
+  testTypeSerde();
+}
+
+TEST_F(OpaqueCustomTypeTest, registerWithSerde) {
+  // Register type with serialization/deserialization and build a TypePtr.
+  TestCustomTypeRegister::registerType(
+      /*serializeValueFunc=*/[](const std::shared_ptr<
+                                 TestCustomType>&) { return ""; },
+      /*deserializeValueFunc=*/
+      [](const std::string&) { return std::make_shared<TestCustomType>(); });
+
+  testTypeSerde();
```

Comment:
Can we verifiy that serializeValueFunc and deserializeValueFunc were actually registered and are functional?